verilog/analysis/checkers/always_ff_non_blocking_rule.cc

```diff
@@ -48,50 +51,150 @@ absl::string_view AlwaysFFNonBlockingRule::Name() {
+  return LintRuleStatus(violations_, Name(), GetStyleGuideCitation(kTopic));
 }
 
-void AlwaysFFNonBlockingRule::HandleSymbol(const verible::Symbol& symbol,
-                                           const SyntaxTreeContext& context) {
-  verible::matcher::BoundSymbolManager manager;
-
-  if (AlwaysFFMatcher().Matches(symbol, &manager)) {
-    for (const auto& match :
-         SearchSyntaxTree(symbol, NodekNetVariableAssignment())) {
-      // this is intended to ignore assignments in for loop step statements like
-      // i=i+1
-      if (match.context.IsInside(NodeEnum::kLoopHeader)) continue;
-
-      const auto* node =
-          dynamic_cast<const verible::SyntaxTreeNode*>(match.match);
-
-      if (node == nullptr) continue;
-
-      const auto* leaf = dynamic_cast<const verible::SyntaxTreeLeaf*>(
-          node->children()[1].get());
-
-      if (leaf == nullptr) continue;
+//- Configuration -----------------------------------------------------------
+absl::Status AlwaysFFNonBlockingRule::Configure(
+    const absl::string_view configuration) {
+  using verible::config::SetBool;
+  return verible::ParseNameValues(
+      configuration, {
+                         {"catch_modifying_assignments",
+                          SetBool(&catch_modifying_assignments_)},
+                         {"waive_for_locals", SetBool(&waive_for_locals_)},
+                     });
+}
 
-      if (leaf->get().token_enum() == '=')
-        violations_.insert(LintViolation(*leaf, kMessage, match.context));
+//- Processing --------------------------------------------------------------
+void AlwaysFFNonBlockingRule::HandleSymbol(const verible::Symbol &symbol,
+                                           const SyntaxTreeContext &context) {
+  static const Matcher always_ff_matcher{
+      NodekAlwaysStatement(AlwaysFFKeyword())};
+  static const Matcher block_matcher{NodekBlockItemStatementList()};
+  static const Matcher decl_matcher{NodekDataDeclaration()};
+  static const Matcher var_matcher{NodekRegisterVariable()};
+  static const Matcher asgn_blocking_matcher{NodekNetVariableAssignment()};
+  static const Matcher asgn_modify_matcher{NodekAssignModifyStatement()};
+  static const Matcher asgn_incdec_matcher{NodekIncrementDecrementExpression()};
+  static const Matcher ident_matcher{NodekUnqualifiedId()};
+
+  // Determine depth in syntax tree and discard state from branches already left
+  const int depth = context.size();
+  if (depth <= inside_) inside_ = 0;
+  while (depth <= scopes_.top().first) {
+    scopes_.pop();
+    VLOG(4) << "POPped to scope DEPTH=" << scopes_.top().first
+            << "; #locals_=" << scopes_.top().second << std::endl;
+  }
+  locals_.resize(scopes_.top().second);
+
+  verible::matcher::BoundSymbolManager symbol_man;
+
+  if (always_ff_matcher.Matches(symbol, &symbol_man)) {
+    // Check for entering an always_ff block
+    VLOG(4) << "always_ff @DEPTH=" << depth << std::endl;
+    inside_ = depth;
+  } else if (inside_) {
```

Comment:
This `else if (inside_)` goes all the way to the end but it increases the mental burden of keeping that in mind. So I suggest to do something like

```c++
//...
else if (!inside_) {
   return
}
```

Then the lower chain of if/else's is easier to follow.

But even then, things are pretty nested below. Maybe we should have one toplevel if/else and then call into separate private methods ? That would know down the cyclomatic complexity level one more.

```c++
if (block_matcher.Matches(symbol, &symbol_man)) {
    scopes.emplace(depth, locals_.size());
} else if (decl_matcher.Matches(symbol, &symbol_man)) {
   CollectLocalVariableDeclarations(...);
} else if (!context.IsInside(NodeEnum::kLoopHeader)) {
   CheckBlockingAssignmentsOutsieLoopheaders(...);
}
...
```